File: cpp/Dsu.cpp
#include <vector>
#include <numeric>
#include <algorithm>
struct Dsu
{
    int n;
    std::vector<int> f, sz;
    Dsu(int _n) : n(_n)
    {
        f.resize(n);
        sz.assign(n, 1);
        std::iota(f.begin(), f.end(), 0);
    }
    int leader(int x)
    {
        if (f[x] == x)
        {
            return x;
        }
        return f[x] = leader(f[x]);
    }
    bool merge(int u, int v)
    {
        u = leader(u), v = leader(v);
        if (u == v)
        {
            return false;
        }
        if (sz[u] < sz[v])
        {
            std::swap(u, v);
        }
        f[v] = u;
        sz[u] += sz[v];
        return true;
    }
    bool same(int u, int v)
    {
        return leader(u) == leader(v);
    }
    int size(int x)
    {
        return sz[leader(x)];
    }
};

File: python/Dsu.py
class Dsu:
    def __init__(self, n: int) -> None:
        self.f = list(range(n))
        self.sz = [1]*n

    def leader(self, x: int) -> int:
        x_copy = x
        while x != self.f[x]:
            x = self.f[x]
        while x_copy != x:
            self.f[x_copy], x_copy = x, self.f[x_copy]
        return x

    def same(self, u: int, v: int) -> bool:
        return self.leader(u) == self.leader(v)

    def merge(self, u: int, v: int) -> bool:
        u, v = self.leader(u), self.leader(v)
        if u == v:
            return False
        self.f[v] = u
        self.sz[u] += self.sz[v]
        return True
